src/bidiMapper/modules/context/NavigationTracker.ts

```diff
@@ -400,64 +411,48 @@ export class NavigationTracker {
     }
 
     navigation.isFragmentNavigation = loaderId === undefined;
+  }
 
-    if (loaderId === undefined || this.#currentNavigation === navigation) {
-      // If the command's navigation is same-document or is already the current one,
-      // nothing to do.
-      return;
-    }
-
-    this.#currentNavigation.fail(
-      'navigation canceled by concurrent navigation',
-    );
-
-    navigation.start();
-    this.#currentNavigation = navigation;
+  frameStartedNavigating(
+    url: string,
+    loaderId: string,
+    navigationType: string,
+  ) {
+    this.#logger?.(LogType.debug, `frameStartedNavigating ${url}, ${loaderId}`);
 
-    if (this.#pendingNavigation === navigation) {
+    if (![loaderId, undefined].includes(this.#pendingNavigation?.loaderId)) {
+      // If there is a pending navigation with loader id set, but not equal to the new
+      // loader id, cancel pending navigation.
+      this.#pendingNavigation?.fail(
+        'navigation canceled by concurrent navigation',
+      );
       this.#pendingNavigation = undefined;
     }
-  }
-
-  /**
-   * Emulated event, tight to `Network.requestWillBeSent`.
-   */
-  frameStartedNavigating(url: string, loaderId: string) {
-    this.#logger?.(LogType.debug, `frameStartedNavigating ${url}, ${loaderId}`);
 
     if (this.#loaderIdToNavigationsMap.has(loaderId)) {
-      // The `frameStartedNavigating` is tight to the `Network.requestWillBeSent` event
-      // which can be emitted several times, e.g. in case of redirection. Nothing to do in
-      // such a case.
+      const existingNavigation = this.#loaderIdToNavigationsMap.get(loaderId)!;
+      // Navigation can be changed from `sameDocument` to `differentDocument`.
+      existingNavigation.isFragmentNavigation = [
+        'historySameDocument',
+        'sameDocument',
+      ].includes(navigationType);
+      this.#pendingNavigation = existingNavigation;
       return;
     }
 
     const pendingNavigation =
       this.#pendingNavigation ?? this.createPendingNavigation(url, true);
 
-    pendingNavigation.url = url;
-    pendingNavigation.start();
-
-    pendingNavigation.loaderId = loaderId;
     this.#loaderIdToNavigationsMap.set(loaderId, pendingNavigation);
-  }
 
-  /**
-   * In case of `beforeunload` handler, the pending navigation should be marked as started
-   * for consistency, as the `browsingContext.navigationStarted` should be emitted before
-   * user prompt.
-   */
-  beforeunload() {
-    this.#logger?.(LogType.debug, `beforeunload`);
+    pendingNavigation.isFragmentNavigation = [
+      'historySameDocument',
+      'sameDocument',
+    ].includes(navigationType);
```

Comment:
Can we extract this check and reuse it we always make this check for both exiting and pending navigation.

Reply:
done